lib/merkle_tree.ex

```diff
@@ -42,35 +45,55 @@ defmodule MerkleTree do
     Alternatively, you can supply your own hash function that has the spec
     ``(String.t -> String.t)``.
   """
-  @spec new(blocks, hash_function) :: t
-  def new(blocks, hash_function \\ &MerkleTree.Crypto.sha256/1)
-  when blocks != [] do
-    unless is_power_of_n(@number_of_children, Enum.count(blocks)), do: raise MerkleTree.ArgumentError
-
-    root = build(blocks, hash_function)
+  @spec new(blocks, hash_function, height) :: t
+  def new(blocks, hash_function \\ &MerkleTree.Crypto.sha256/1, height \\ @default_height) do
+    root = build(blocks, hash_function, height)
+    blocks = blocks |> extend_with_zeroes(height)
     %MerkleTree{blocks: blocks, hash_function: hash_function, root: root}
   end
 
+  defp extend_with_zeroes(blocks, height) do
+    extension_length = pow(height) - Enum.count(blocks)
+    blocks ++ List.duplicate(@zeroes, extension_length)
+  end
+
+  defp pow(n), do: :math.pow(2, n) |> round
+
   @doc """
     Builds a new binary merkle tree.
   """
-  @spec build(blocks, hash_function) :: root
-  def build(blocks, hash_function) do
-    starting_height = 0
+  def build(blocks, hash_function, max_height) when blocks != [] do
```

Comment:
when doing recursive calls, place base case (`def build([],...)`) at the top

Reply:
done